### ISSUE TITLE
dataclass_transform decorator causes class-level descriptor access to not honor the descriptor

### ISSUE THREAD

**Issue by zzzeek** [microsoft/pyright#3486]
When a descriptor returns an alternate kind of object at the class level, typing tools such as Pyright correctly honor this.   However when a decorator that itself is decorated with `@dataclass_transform` is applied to the class, this behavior breaks and descriptor access appears to be hardcoded to returning the type of the descriptor itself.  This seems to occur regardless of whether or not that descriptor is one of the `field_descriptors` for the transform:

```py
from __future__ import annotations
from dataclasses import dataclass

from typing import (
    overload,
    Any,
    TypeVar,
    Generic,
    Optional,
    Union,
    Callable,
    Type,
)
from typing_extensions import dataclass_transform


_T = TypeVar("_T")
_O = TypeVar("_O")


class ClassLevelSQLExpression(Generic[_T]):
    """some kind of SQL-oriented construct"""


class Mapped(Generic[_T]):
    """a descriptor which returns a SQL composition object at the class level"""

    @overload
    def __get__(
        self, instance: None, owner: Any
    ) -> ClassLevelSQLExpression[_T]:
        ...

    @overload
    def __get__(self, instance: object, owner: Any) -> _T:
        ...

    def __get__(
        self, instance: Optional[object], owner: Any
    ) -> Union[ClassLevelSQLExpression[_T], _T]:
        ...


def mapped_normally() -> Callable[[Type[_O]], Type[_O]]:
    ...

# behavior is the same regardless of field_descriptors being present
@dataclass_transform(field_descriptors=(Mapped[Any],))
def mapped_as_dataclass(
    *,
    init: bool = True,
    repr: bool = True,  # noqa: A002
    eq: bool = True,
    order: bool = False,
    unsafe_hash: bool = False,
) -> Callable[[Type[_O]], Type[_O]]:
    ...


@mapped_normally()
class NormalMappedClass:
    id: Mapped[int]
    data: Mapped[str]


@mapped_as_dataclass()
class DataclassMappedClass:
    id: Mapped[int]
    data: Mapped[str]
    data_w_assign: Mapped[str] = Mapped()



@dataclass
class RealDataclass:
    id: Mapped[int]
    data: Mapped[str]
    data_w_assign: Mapped[str] = Mapped()


reveal_type(NormalMappedClass.data)

reveal_type(DataclassMappedClass.data)

reveal_type(DataclassMappedClass.data_w_assign)

reveal_type(RealDataclass.data)

reveal_type(RealDataclass.data_w_assign)



```

pyright output:

```
/home/classic/dev/sqlalchemy/test3.py
  /home/classic/dev/sqlalchemy/test3.py:81:13 - information: Type of "NormalMappedClass.data" is "ClassLevelSQLExpression[str]"
  /home/classic/dev/sqlalchemy/test3.py:83:13 - information: Type of "DataclassMappedClass.data" is "Mapped[str]"
  /home/classic/dev/sqlalchemy/test3.py:85:13 - information: Type of "DataclassMappedClass.data_w_assign" is "Mapped[str]"
  /home/classic/dev/sqlalchemy/test3.py:87:13 - information: Type of "RealDataclass.data" is "Mapped[str]"
  /home/classic/dev/sqlalchemy/test3.py:89:13 - information: Type of "RealDataclass.data_w_assign" is "Mapped[str]"
0 errors, 0 warnings, 5 informations 
```

```
$ pyright --version
pyright 1.1.247

```

**Comment by zzzeek** [comment]
It appears that pyright reports the same information for a real dataclass also.

Mypy, OTOH, reports ``test3.ClassLevelSQLExpression[builtins.str]`` for all answers, including a real dataclass:

```py
from dataclasses import dataclass


@dataclass
class RealDataclass:
    id: Mapped[int]
    data: Mapped[str]
    data_w_assign: Mapped[str] = Mapped()

reveal_type(RealDataclass.data)

reveal_type(RealDataclass.data_w_assign)

```

mypy:
```
test3.py:87: note: Revealed type is "test3.ClassLevelSQLExpression[builtins.str]"
test3.py:89: note: Revealed type is "test3.ClassLevelSQLExpression[builtins.str]"
```

I'm fairly nervous about this, because all of these behaviors determine whether or not I can proceed with my dataclasses integration, and pep-681 doesn't have any language indicating how these things should behave.

**Comment by zzzeek** [comment]
Here's an example script that illustrates using traditional `@dataclass`, the descriptor's class-level type is honored at runtime, whether or not there's an explicit assignment on the class itself:


```py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any
from typing import Generic
from typing import Optional
from typing import overload
from typing import TYPE_CHECKING
from typing import TypeVar
from typing import Union


_T = TypeVar("_T")
_O = TypeVar("_O")


class ClassLevelSQLExpression(Generic[_T]):
    """some kind of SQL-oriented construct"""


class Mapped(Generic[_T]):
    """a descriptor which returns a SQL composition object at the class level"""

    @overload
    def __get__(
        self, instance: None, owner: Any
    ) -> ClassLevelSQLExpression[_T]:
        ...

    @overload
    def __get__(self, instance: object, owner: Any) -> _T:
        ...

    def __get__(
        self, instance: Optional[object], owner: Any
    ) -> Union[ClassLevelSQLExpression[_T], _T]:
        if instance is None:
            return ClassLevelSQLExpression()
        else:
            raise NotImplementedError()

class Base:
    data = Mapped()

@dataclass
class RealDataclass(Base):
    id: Mapped[int]
    data: Mapped[str]
    data_w_assign: Mapped[str] = Mapped()


assert isinstance(RealDataclass.data, ClassLevelSQLExpression)
assert isinstance(RealDataclass.data_w_assign, ClassLevelSQLExpression)

if TYPE_CHECKING:
    reveal_type(RealDataclass.data)
    reveal_type(RealDataclass.data_w_assign)
```

**Comment by erictraut** [comment]
Is your intent for fields `id`, `data` and `data_w_assign` to have corresponding parameters in the synthesized `__init__` method? If so, what type do you expect those parameters to have?

If your intent is not for these fields to have corresponding parameters in the synthesized `__init__`, why are you not marking them as `ClassVar`s? If you annotate them as `ClassVar`s, you will see the behavior you're expecting when accessing them from the class.

**Comment by zzzeek** [comment]
> Is your intent for fields `id`, `data` and `data_w_assign` to have corresponding parameters in the synthesized `__init__` method? 

yes.

>  If so, what type do you expect those parameters to have?

as discussed in detail at https://github.com/microsoft/pyright/discussions/2958# as well as on a Zoom call with both you and @debonte  , the types of parameters would be the "_T" type that is expressed as the type received by the ``__set__`` method on the descriptors.  I realize the ``__set__`` method isn't present in my example above which I omitted for brevity, however if I restore that method above, the outcome is the same as the issue is only dealing with the classlevel behavior of the descriptor.


> 
> If your intent is not for these fields to have corresponding parameters in the synthesized `__init__`, why are you not marking them as `ClassVar`s? 

because it is not necessary from a typing perspective, and as discussed in previous conversations, our syntax is one where users will be typing out `Mapped[]`, and they should not have to wrap it in a redundant `ClassVar` declaration, else I would have architected this differently in order to accommodate such a limitation.  Typing tools don't require descriptors to be wrapped in `ClassVar` in order to function correctly at the class level.  

If you recall, we had originally discussed the feasibility of including a parameter called `transform_descriptor_types` in pep681, but it was removed when it was realized that the existing behavior of dataclasses is such that descriptors already do exactly what I need, which is absolutely true, and as it was agreed that the spirit of pep681 is to duplicate all behaviors of dataclasses as much as possible, such a parameter would not be needed.

So here I am noting that pyright's current behavior is not consistent with the behavior of dataclasses.  When dataclass_transform was first available in pyright, I tested everything out and I would be surprised if this wasn't working at that time so this may be a regression.


> If you annotate them as `ClassVar`s, you will see the behavior you're expecting when accessing them from the class.

To the extent that I could perhaps get the "Mapped" symbol to actually be an alias to ClassVar[] to a descriptor might be a workaround, but I'm not really sure if that would work, but overall I shouldn't have to because typing tools honor descriptors without ClassVar being present.  The presence of `@dataclass` doesn't actually change that at runtime.

**Comment by zzzeek** [comment]
Adding ClassVar leads pyright to no longer consider the descriptor's name and `__set__` datatype as part of the generated `__init__` method, so ClassVar can't work.  I also can't make a type alias with ClassVar that has a TypeVar inside of it so workarounds not feasible.   In any case, I still observe the broken class behavior when adding ClassVar, see demonstration below

```py
from __future__ import annotations

from typing import Any
from typing import Callable
from typing import ClassVar
from typing import Generic
from typing import Optional
from typing import overload
from typing import Type
from typing import TypeVar
from typing import Union

from typing_extensions import dataclass_transform


_T = TypeVar("_T")
_O = TypeVar("_O")


class ClassLevelSQLExpression(Generic[_T]):
    """some kind of SQL-oriented construct"""


class Mapped(Generic[_T]):
    """a descriptor which returns a SQL composition object at the class level"""

    @overload
    def __get__(
        self, instance: None, owner: Any
    ) -> ClassLevelSQLExpression[_T]:
        ...

    @overload
    def __get__(self, instance: object, owner: Any) -> _T:
        ...

    def __get__(
        self, instance: Optional[object], owner: Any
    ) -> Union[ClassLevelSQLExpression[_T], _T]:
        ...


    def __set__(
        self, instance: Any, value: Union[ClassLevelSQLExpression[_T], _T]
    ) -> None:
        ...

    def __delete__(self, instance: Any) -> None:
        ...


@dataclass_transform(field_descriptors=(Mapped[Any],))
def mapped_as_dataclass(
    *,
    init: bool = True,
    repr: bool = True,  # noqa: A002
    eq: bool = True,
    order: bool = False,
    unsafe_hash: bool = False,
) -> Callable[[Type[_O]], Type[_O]]:
    ...



@mapped_as_dataclass()
class DataclassMappedClass:
    id: Mapped[int]
    data: Mapped[str]
    data_w_assign: Mapped[str] = Mapped()
    data_w_classvar: ClassVar[Mapped[str]] = Mapped()

reveal_type(DataclassMappedClass.data)

reveal_type(DataclassMappedClass.data_w_assign)

reveal_type(DataclassMappedClass.data_w_classvar)


DataclassMappedClass(id=1,data='d', data_w_assign='dd', data_w_classvar='ddd')
```

output:

```
 /home/classic/dev/sqlalchemy/test3.py:72:13 - information: Type of "DataclassMappedClass.data" is "Mapped[str]"
  /home/classic/dev/sqlalchemy/test3.py:74:13 - information: Type of "DataclassMappedClass.data_w_assign" is "Mapped[str]"
  /home/classic/dev/sqlalchemy/test3.py:76:13 - information: Type of "DataclassMappedClass.data_w_classvar" is "Mapped[str]"
  /home/classic/dev/sqlalchemy/test3.py:79:57 - error: No parameter named "data_w_classvar" (reportGeneralTypeIssues)
1 error, 0 warnings, 3 informations 

```

**Comment by erictraut** [comment]
Yes, if your intent is for these variables to be part of `__init__`, then `ClassVar` isn't appropriate. That's why I asked what your intent was.

**Comment by erictraut** [comment]
This will be fixed in the next release. The fix applies to both dataclass and dataclass_transform.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.248, which I just published. It will also be included in the next release of pylance.

**Comment by zzzeek** [comment]
certainly not at runtime because when we make the dataclass, the "descriptor" that we use is not actually there on the class at that time; for both our old and new "dataclasses" usage patterns, when `@dataclass` gets the class, it sees `dataclass.Field` objects in all those places.

at typing time, I think we are OK because the configurational "descriptor" objects that are applied to the class in source code are all passed as field_specifiers to dataclass_transform, so any behaviors where default values for fields are derived from the "descriptor" object should not take place, as the object present is a field specifier with a "default" and "default_factory" field it should be looking for.
